CIP-array-builtin/README.md

```diff
@@ -0,0 +1,296 @@
+We considered the following options:
+
+1. Introduce another new builtin type, one which implements `Maybe` semantics. The type
+signature for `indexArray` would become `forall a . Integer -> Array a -> Maybe a` and it
+would return `Nothing` for out-of-bounds lookups.
+    - The obvious disadvantage is the necessity of adding another new builtin type (there is
+    no `Maybe` builtin type in Plutus Core), which would further increase the complexity of
+    the builtin language.
+    - Another disadvantage would be that this solution is the most costly: users will incur
+    additional costs in deconstructing the returned `Maybe`.
+
+2. Failed lookups return a default value provided by the caller:
+`indexArray :: forall a . Integer -> a -> Array a -> a`.
+    - This solution is problematic whenever there is no sensible default value and
+    the user wants the function to fail. Since Plutus Core is strict, it is not possible
+    to pass `error` as the default value without it getting evaluated before the call and
+    terminating execution immediately.
+
+3. Include `lengthOfArray` and require the user to perform appropriate length guards before
+calling `indexArray`.
+    - This is a familiar option from other languages.
+    - By definition arrays are fixed-size and their length is available in constant time.
+    - It also allows users to omit bounds checks when they know a priori that the
+    index is in bounds.
+
+After considering these three options, we concluded that the addition of
+the `lengthOfArray` builtin function is both necessary and sufficient for introducing a
+well-defined array type into the builtin language.
+
+### Constructing arrays
+
+The last required functionality for having a practical interface is the ability to
+construct arrays.
+
+Due to our decision of providing immutable arrays, it is difficult (or very expensive) to build up arrays incrementally.
+A naive approach would require repeated copying and potentially the usage of quadratic
+space and time.
+
+A more appropriate approach would be to construct the array in bulk,
+however that would require an intermediate representation of a collection of elements. Fortunately this already exists in the builtin language in the form of builtin lists.
+We can then naturally introduce a function which transforms lists into arrays: `listToArray`.
+
+Since the only current method of constructing arrays is via `listToArray`, one may ask
+whether the `lengthOfArray` builtin is at all necessary.
+We believe that, indeed, `lengthOfArray` is redundant in the current specification. The reason is
+that traversing the list to construct the array is enough to also compute its length. 
+However, we have decided that `lengthOfArray` should be kept as future work will depend upon its
+existence. Some components of Plutus Core could be natively represented as arrays,
+and there would be no other way to access their length in constant time other than
+through a builtin.
```

Comment:
Doesn't sound very convincing. We know that `dropList` for example is gonna definitely help some and it's still not added. But if `lengthOfArray` is described as "kept as future work", it just doesn't sound like a compelling argument within our current "builtins tradition".

I personally think that `lengthOfArray` is gonna be useful, because people will hate to pass the additional length parameter for array-polymorphic functions or even regular ones, calling `lengthOfArray` is easier than allocating a variable for the length, remembering what it was and referencing at each call site. Since arrays are static in a lot of cases we'll be able to optimize `lengthOfArray` calls away anyway.

Reply:
Yeah, and I think the point about receiving arrays of unknown length is pretty fatal to this idea.

Reply:
I completely removed this section because it's clear we will need to try the "future work" as part of this CIP.

Reply:
@rphair please resolve this comment.